job-runner/chain/reaper.go

```diff
@@ -0,0 +1,589 @@
+			return
+		default:
+			// If there were no jobs done running, and there are no jobs currently
+			// running, then there must be no more jobs to run - the chain is done.
+			if len(r.running) == 0 {
+				break REAPER
+			}
+		}
+	}
+
+	r.Finalize()
+	close(r.doneChan)
+}
+
+// Stop stops the reaper from reaping any more jobs. It blocks until the reaper
+// is stopped (will reap no more jobs and Run will return).
+func (r *RunningChainReaper) Stop() {
+	select {
+	case <-r.stopChan:
+	default:
+		close(r.stopChan)
+	}
+
+	<-r.doneStopping
+	return
+}
+
+// reap takes a job that just finished running, saves its final state, and prepares
+// to continue running the chain (or recognizes that the chain is done running).
+//
+// If chain is done: save final state + stop running more jobs.
+// If job failed:    retry sequence if possible.
+// If job completed: prepared subsequent jobs and enqueue if runnable.
+func (r *RunningChainReaper) Reap(job proto.Job) {
+	jLogger := r.logger.WithFields(log.Fields{"job_id": job.Id, "sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+
+	// Set the final state of the job in the chain.
+	r.chain.SetJobState(job.Id, job.State)
+	r.chainRepo.Set(r.chain)
+
+	if job.State == proto.STATE_FAIL {
+		if r.chain.CanRetrySequence(job.Id) {
+			// Retry the sequence if possible.
+			jLogger.Info("job did not complete successfully. retrying sequence.")
+			r.prepareSequenceRetry(job)
+
+			// Re-enqueue the first job of the sequence and increment the
+			// sequence's try count. Subsequent jobs in the sequence will be
+			// enqueued after the first job completes.
+			sequenceStartJob := r.chain.SequenceStartJob(job.Id)
+			r.chain.IncrementSequenceTries(sequenceStartJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+			r.send(sequenceStartJob) // re-enqueue first job in sequence
+			return
+		}
+
+		// If the failed job was stopped or is not part of a retryable
+		// sequence, then ignore subsequent jobs. For example, if job B of
+		// A -> B -> C  fails, then C is not run.
+		jLogger.Warn("job did not complete successfully.")
+		return
+	}
+	jLogger.Infof("job completed successfully.")
+
+	// Copy job data from this job to all child jobs. When a job has multiple
+	// parent jobs, it will get jobData copied from all of its parent jobs in
+	// the order that they finish. Therefore, if multiple parent jobs try to set
+	// the same jobData field, it is undetermined which parent job the job will
+	// receive the final copy of that jobData field from.
+	for _, nextJob := range r.chain.NextJobs(job.Id) {
+		nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJob.Id})
+
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
```

Comment:
Do this first, i.e. don't copy data and call chainRepo.Set then skip the job because it's not runnable.

Reply:
The idea behind this was to copy the data of all parent jobs to each child job, not just the data from the last parent job to finish.

Reply:
Code comment :-)